### PR TITLE
Document and comment some of the iterator code

### PR DESCRIPTION
**Pull request by nothingmuch**
Caveat: Low effort / noisy PR, but seems justifiable.

Since the code is a bit hard to follow, in trying to evaluate it I ended up adding some comments. Also includes sweeping clippy fixes, on the one hand those improve readability bot on the other it's arguably an unrelated change and touches many more files than the ones I commented, LMK if you'd prefer that to be less hamfisted (e.g. only affect touched files).

There is potentially a bug which can't be triggered, added a FIXME comment for that just for the sake of posterity, though I have no intention on fixing this.

Roughly estimating the complexity led me to conclude that this code could be much improved, potentially making the performance reasonable for slightly larger transactions (not much larger since the overall behavior is still exponential).

### CODE REVIEW

File: src/partition/mod.rs [hunk 1]
```diff
@@ -43,14 +50,28 @@ impl<'a> SumFilteredPartitionIterator<'a> {
         }
     }
 
+    /// Complexity is $O(\mathrm{FML})$, also in the average case.
```
Comment (DanGould) [inline src/partition/mod.rs]: What do you mean by $O(\mathrm{FML})$? "Wowo this is so high I should just give up?" Filter, Memory, Length?

File: src/partition/mod.rs [hunk 2]
```diff
@@ -72,21 +99,38 @@ impl<'a> SumFilteredPartitionIterator<'a> {
...
         let next = match self.right_partitions_iterator {
             Some(ref mut iter) => iter.next(),
             None => None,
         };
         match next {
+            // Since the nested partition iterator shares the same filter is
+            // this one, all parts of the partition of the complement (right)
+            // set match, as does the subset (left set). Extend the partition
+            // with the left set as a part, and return the resulting partition
+            // of the initial set.
```
Comment (DanGould) [inline src/partition/mod.rs]:
```suggestion
            // Since the nested partition iterator shares the same filter as
            // this one, all parts of the partition of the complement (right)
            // set match, as does the subset (left set). Extend the partition
            // with the left set as a part, and return the resulting partition
            // of the initial set.
```

**Review by DanGould — APPROVED** [review]
Which cargo version did you use to "make clippy happy?" I'm still getting clippy errors with 1.60.0 (this crate's specified `rust-version`) and 1.63 (MSRV for most of the rust-bitcoin ecosystem), stable, and nightly on [e1ccd51](https://github.com/payjoin/cja/pull/1/commits/e1ccd510e6d0059db4ae3625c4c8ecc27b19c706). Perhaps the intention was just to make clippy *happier* which this commit does.

File: src/partition/mod.rs (hunk 1 above)
Comment (nothingmuch) [reply src/partition/mod.rs]: oh i was just too lazy/tired to work it out and i thought i was being funny, that's not a product of terms but an acronym =P

asymptotically it's exponential, i think with constant factor in the exponent and more precisely another linear term so something like `O(n*2^{k*n})` except that's not the meaning of big O anymore but accounting for the bloom filters gating some of that (i.e. what is the expected negative rate). it doesn't really matter what it is theoretically though

Comment (DanGould) [reply src/partition/mod.rs]: I'd understand if it were in plain language saying ~"Exponential, hard to calculate exactly, doesn't really matter"

**Comment by nothingmuch** [comment]
yesterday I didn't notice but forgot to push one of the commits.

also rephrased my silly comment to be a bit more coherent

**Comment by nothingmuch** [comment]
> Which cargo version did you use to "make clippy happy?" I'm still getting clippy errors with 1.60.0 (this crate's specified `rust-version`) and 1.63 (MSRV for most of the rust-bitcoin ecosystem), stable, and nightly on [e1ccd51](https://github.com/payjoin/cja/pull/1/commits/e1ccd510e6d0059db4ae3625c4c8ecc27b19c706). Perhaps the intention was just to make clippy _happier_ which this commit does.

Yes, make clippy _happier_ is definitely more precise, but even that is a bit of a reach because it's just `clippy --fix`, i'll revise the commit message

cargo clippy --version:
clippy 0.1.77 (25ef9e3 2024-04-09)

**Review by DanGould — APPROVED** [review]
